sequencer/src/main/java/net/consensys/linea/bundles/BundleForwarder.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Consensys Software Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package net.consensys.linea.bundles;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.OptionalLong;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+import net.consensys.linea.bundles.BundlePoolService.TransactionBundleAddedListener;
+import net.consensys.linea.bundles.BundlePoolService.TransactionBundleRemovedListener;
+import net.consensys.linea.utils.PriorityThreadPoolExecutor;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.hyperledger.besu.plugin.services.BlockchainService;
+
+@Slf4j
+@RequiredArgsConstructor
+class BundleForwarder implements TransactionBundleAddedListener, TransactionBundleRemovedListener {
+  private final AtomicLong reqIdProvider = new AtomicLong(0L);
+  private final PriorityThreadPoolExecutor executor;
+  private final BlockchainService blockchainService;
+  private final OkHttpClient rpcClient;
+  private final URL recipientUrl;
+
+  @Override
+  public void onTransactionBundleAdded(final TransactionBundle bundle) {
+    executor.submit(new SendBundleTask(bundle, 0));
+  }
+
+  @Override
+  public void onTransactionBundleRemoved(final TransactionBundle transactionBundle) {
+    executor.remove(new SendBundleTask(transactionBundle, 0));
+  }
+
+  void retry(final TransactionBundle bundle, final int retry) {
+    executor.submit(new SendBundleTask(bundle, retry));
```

Comment:
I was expecting a max retry limit here, just to keep the queue manageable if a host goes offline.  Are you thinking that retries will naturally be dropped as their priority value gets too high

Reply:
Since a bundle naturally expires when its block number is imported, I am using that as a _retry limit_, basically a bundle will be retried until it is still valid according to its block number.

If this strategy will not be good enough, then we can evaluate if we need a strict retry limit.

Reply:
[As described below ](https://github.com/Consensys/linea-sequencer/pull/178#discussion_r2009991831) the upper bound is driven by the limits of the cache, so there could not be more queue tasks than bundle in the cache.

Reply:
ah ok - that is the missing bit.   LimitedBundlePool has a max weight (not in the diff) 👍